### ISSUE TITLE
How to find mAP (mean absolute precision) of tflite of object detection?

### ISSUE THREAD

**Issue by MhmudAlpurd** [ultralytics/yolov5#5481]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi everyone, I converted the PyTorch file of yolov5 to tflite (object detection model). and I’m going to get the mAP(mean absolute precision) or other performance metrics of this file. but I can’t find code or a way to do that. do you have this experience or solution?

thank you

**Comment by glenn-jocher** [comment]
@MhmudAlpurd there are no existing methods for obtaining mAP on exported models. You can run inference using detect.py on many model formats including TFLite for anecdotal comparisons however.

**Comment by MhmudAlpurd** [comment]
@glenn-jocher  thanks a lot for your helping.

**Comment by glenn-jocher** [comment]
@MhmudAlpurd a number of users have asked about this, so we'll likely do a PR in the future to add this capability, but be aware TFLite inference is typically very slow on CPU, so running val.py on COCO for example with a TFLite model will take a very long time.

**Comment by glenn-jocher** [comment]
@MhmudAlpurd good news 😃! Your original issue may now be fixed ✅ in PR #5549. This PR consolidates all backends into a single YOLOv5 `DetectMultiBackend()` class. Supports following model formats:

- [x] PyTorch `*.pt`
- [x] TorchScript `*.torchscript.pt`
- [x] CoreML `*.mlmodel`
- [x] TensorFlow `saved_model`
- [x] TensorFlow `*.pb`
- [x] TensorFlow Lite `*.tflite`
- [ ] TensorFlow.js `web_model`
- [x] ONNX Runtime `*.onnx`
- [x] OpenCV DNN `*.onnx`


### Usage
```python
# Export
python export.py --weights yolov5s.pt --include tflite

# Inference
python detect.py --weights yolov5s.tflite
python val.py --weights yolov5s.tflite
```

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by MhmudAlpurd** [comment]
@glenn-jocher you are amazing. thanks a lot.

**Comment by Hyungjun-K1m** [comment]
@glenn-jocher It's amazing that the library already supports tflite inference!

However, it seems like the tflite model (https://github.com/ultralytics/yolov5/releases/download/v6.0/yolov5s.tflite) shows much lower mAP than the baseline model.

I'd like to know if there was any mistake.

I used the Raspberry Pi 4 B device with 64-bit linux OS and run
`python val.py --data coco.yaml --weights './yolov5s.tflite' --imgsz 640`

The results were

|                 	| mAP@.5 	| mAP@.5:.95 	| 
|:---------------:	|:------:	|:----------:	|
|    local mAP    	|  0.512 	|    0.301   	|
| pycocotools mAP 	|  0.519 	|    0.315   	|


While the baseline mAP (pycocotools) results are 0.564 for mAP@.5 and 0.372 for mAP@.5:.95, the tflite results are qutie low.
It might be due to the INT8 quantization of tflite conversion, but it's much severe than I expected.

Do you think this result is correct result?

Thanks!

**Comment by glenn-jocher** [comment]
@Hyungjun-K1m TFLite and CoreML exported models are at --img 320 image size suitable for mobile deployments. The PyTorch models are trained at --img 640.

If you export a tflite model yourself you will get the same mAP:

```bash
python export.py --weights yolov5s.pt --include tflite

python val.py --weights yolov5s.pt
python val.py --weights yolov5s-fp16.tflite
```

**Comment by glenn-jocher** [comment]
<img width="922" alt="Screenshot 2021-12-06 at 16 19 19" src="https://user-images.githubusercontent.com/26833433/144872171-b8067777-03a5-4731-96ab-9588bbbee785.png">

**Comment by Hyungjun-K1m** [comment]
![yolov5s_head](https://user-images.githubusercontent.com/40038794/144952991-1fde9dee-81c8-45f3-9d8f-444f3e507e84.PNG)

@glenn-jocher 
What I see from the downloaded tflite file is that it requires 640x640 inputs. 
Thus I got ValueError about Dimension mismatch (Got 320 but expected 640 for dimension 1 of input 430) when I run the command with --imgsz 320 argument.
What do you mean by 'the tflite models are at --img 320' ?

By the way, I'll try to convert the pt model to tflite model by myself and check the performance.

Thanks for the help!

**Comment by Hyungjun-K1m** [comment]
Unfortunately, the newly exported tflite model showed similar mAP (**0.520 for mAP@.5 and 0.314 for mAP@.5:.95**) as before.
Below command was used to export tflite file.
`python export.py --weights ./yolov5s.pt --img 640 --include tflite --int8`

Considering that the tflite model generated without **--int8** flag shows acceptable mAP (0.561 for mAP@.5 and 0.371 for mAP@.5:.95), I guess the degradation is from int8 ptq process.

Please let me know if you have different opinion! Thanks!

**Comment by glenn-jocher** [comment]
@Hyungjun-K1m yes, for TFLite int8 models I also see about 5% lower mAP than with fp16 models. This is an interesting discovery! The quantization is the only difference so it must be the cause.

**Comment by glenn-jocher** [comment]
@Hyungjun-K1m see https://www.tensorflow.org/lite/performance/post_training_quantization#model_accuracy

It looks like there are also a variety of quantization options, maybe you could experiment with the exact implementation here and let us know if anything works better than the current default:
https://www.tensorflow.org/lite/performance/post_training_quantization

https://github.com/ultralytics/yolov5/blob/554f782537b9af336c02c013468b78fe16ce092d/export.py#L210-L224

**Comment by glenn-jocher** [comment]
@zldrobit interesting discovery today. The YOLOv5s TFLite INT8 models are about 6% lower mAP than the FP16 TFLite/PyTorch models. As of PR https://github.com/ultralytics/yolov5/pull/5845 3 days ago we can now validate all exported model formats.

```bash
python export.py --weights yolov5s.pt --include tflite
python export.py --weights yolov5s.pt --include tflite --int8

python val.py --data coco128.yaml --weights yolov5s.pt  # 41.3 mAP
python val.py --data coco128.yaml --weights yolov5s-fp16.tflite  # 40.9 mAP
python val.py --data coco128.yaml --weights yolov5s-int8.tflite  # 34.7 mAP
```

EDIT: do you think this is this within normal expectations for accuracy loss? I'm not sure myself, but I can try to export CoreML and obtain mAPs to see how well Apple is handling the int8 quantization.

**Comment by zldrobit** [comment]
@glenn-jocher In TFLite, full int8 quantization uses fixed-point integers, and calculates the scale and zero-point of a tensor for quantizing/dequantizing for activations (https://www.tensorflow.org/lite/performance/post_training_quantization#model_accuracy). There are only one scale and one zero-point for a tensor. This requires each value (random variable) in the tensor should have the same distribution, which is not the case. Unlike image classification, object detection needs outputs of object corrdinates. The last output tensor for the TFLite model is of shape `[1, 25200, 5 + #classes]`, and the last dimension consists of `xywh`, objectiveness, and probabilities of classes. Though `xywh` is normalized to `[0, 1]` according to the range of probabilities, the distributions of `xywh` and the rest probabilities are different. And this introduces accuracy drops more or less.

On the other hand, [some experiments on TensorFlow/TFLite](https://arxiv.org/pdf/1806.08342.pdf) (table 3 at the bottom  of page 11) show that mobilenet v1 after post-training quantization (PTQ) has an  accuracy drop about 10%. Changing from PTQ to quantization aware training (QAT) fixes this problem (table 4 at the bottom  of page 21). I excerpt table 4:
![image](https://user-images.githubusercontent.com/3896992/145157540-97a7b480-fb26-4d8c-b396-7207bcdd1843.png)

Generally, if the accuracy after PTQ is acceptable, one could keep using the quantized model. Otherwise, one could (re)train and quantize the model with QAT (fine-tune).

EDIT: another example for QAT advantageous over PTQ https://developer.nvidia.com/blog/improving-int8-accuracy-using-quantization-aware-training-and-tao-toolkit/

**Comment by glenn-jocher** [comment]
@zldrobit super interesting, thanks for the detailed analysis. In TFLite int8 export seems to be applied to the entire model at once, and like you said different layers have different distributions. CoreML actually quantizes each layer individually, applying a range or a lookup table (lookup table based on kmeans), so I think it might handle INT8 better because of this, but it also takes much longer. The CoreML INT8 quantization for a YOLOv5s model takes 5 minutes, and for a YOLOv5x model maybe almost an hour since it runs 256-point kmeans on each layer.

I still haven't had a chance to compare CoreML yet, but I'll try to do it this week. Naturally after #5845 the next logical step is to get mAP for each export format into a table. This is really exciting because we've never been able to validate exported models before, so we are only uncovering some of this now.
